### PR TITLE
Principal::getProvenance can no longer trigger unscheduled execution

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Removed the ability to cause unscheduled execution of an EDProducer be triggered by asking for its Provenance. Only 1 framework test module could actually have caused such behavior.
- Added new Principal::getStableProvenance method which gives access to the provenance information which is stable for the entire job.
- Switched modules to use getStableProvenance if applicable.

#### PR validation:

The code compiles and framework unit tests pass.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Framework/interface/LuminosityBlock.h [hunk 1]
```diff
@@ -136,6 +136,8 @@ namespace edm {
 
     Provenance getProvenance(BranchID const& theID) const;
 
+    StableProvenance getStableProvenance(BranchID const& theID) const;
```
Comment (makortel) [inline FWCore/Framework/interface/LuminosityBlock.h:139]: Should this be 
```suggestion
    StableProvenance const& getStableProvenance(BranchID const& theID) const;
```
as in `Event.h`?

File: FWCore/Framework/interface/Run.h [hunk 2]
```diff
@@ -129,6 +129,8 @@ namespace edm {
 
     Provenance getProvenance(BranchID const& theID) const;
 
+    StableProvenance getStableProvenance(BranchID const& theID) const;
```
Comment (makortel) [inline FWCore/Framework/interface/Run.h:132]: Should this be 
```suggestion
    StableProvenance const& getStableProvenance(BranchID const& theID) const;
```
as in `Event.h`?

File: FWCore/Framework/src/Event.cc [hunk 3]
```diff
@@ -114,12 +114,16 @@ namespace edm {
 
   ProcessHistoryID const& Event::processHistoryID() const { return eventPrincipal().processHistoryID(); }
 
-  Provenance Event::getProvenance(BranchID const& bid) const {
-    return provRecorder_.principal().getProvenance(bid, moduleCallingContext_);
+  Provenance Event::getProvenance(BranchID const& bid) const { return provRecorder_.principal().getProvenance(bid); }
+
+  Provenance Event::getProvenance(ProductID const& pid) const { return eventPrincipal().getProvenance(pid); }
```
Comment (makortel) [inline FWCore/Framework/src/Event.cc:119]: Couldn't these return a const reference as well?
```suggestion
  Provenance const& Event::getProvenance(BranchID const& bid) const { return provRecorder_.principal().getProvenance(bid); }

  Provenance const& Event::getProvenance(ProductID const& pid) const { return eventPrincipal().getProvenance(pid); }
```

File: FWCore/Framework/src/EventPrincipal.cc [hunk 4]
```diff
@@ -351,9 +351,14 @@ namespace edm {
     }
   }
 
-  Provenance EventPrincipal::getProvenance(ProductID const& pid, ModuleCallingContext const* mcc) const {
+  Provenance EventPrincipal::getProvenance(ProductID const& pid) const {
```
Comment (makortel) [inline FWCore/Framework/src/EventPrincipal.cc:354]:
```suggestion
  Provenance const& EventPrincipal::getProvenance(ProductID const& pid) const {
```
?

File: FWCore/Framework/interface/LuminosityBlock.h (hunk 1 above)
Comment (Dr15Jones) [reply FWCore/Framework/interface/LuminosityBlock.h]: done

File: FWCore/Framework/src/Event.cc (hunk 3 above)
Comment (Dr15Jones) [reply FWCore/Framework/src/Event.cc]: Returning a copy appears to have been a left over of when the object used to be created by the call. Changed.

File: FWCore/Framework/src/EventPrincipal.cc (hunk 4 above)
Comment (Dr15Jones) [reply FWCore/Framework/src/EventPrincipal.cc]: done

File: FWCore/Framework/interface/Run.h (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Framework/interface/Run.h]: Thanks for the catch. Done.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+1

**Comment by jfernan2** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32614	25c51b7
- technical changes for the reco code

**Comment by civanch** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by Dr15Jones** [comment]
ping @cms-sw/analysis-l2 @cms-sw/hlt-l2

**Comment by Martin-Grunewald** [comment]
+1

**Comment by silviodonato** [comment]
merge

**Comment by santocch** [comment]
+1

**Pull request merged**